### PR TITLE
Fixes #30903 - introduce at_exit initializer

### PR DESCRIPTION
**Pull request by lzap**
When Prometheus is enabled, it accumulates many temporary files for each worker process and once they get recycled, processes leaves many data file behind making the `/metrics` endpoint slower and slower to the point where it is unusable. This patch introduces `at_exit` block that performs cleanup during shutdown. This will keep Prometheus client library at healthy state.

### CODE REVIEW

**Comment by lzap** [comment]
Failures were unrelated, rebased.

File: config/initializers/at_exit.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
+#
+# Graceful shutdown of web workers
```
Comment (ekohl) [inline config/initializers/at_exit.rb:2]: Puma also has [on_worker_shutdown](https://puma.io/puma/Puma/DSL.html#on_worker_shutdown-instance_method). Would that be more appropriate?

Comment (lzap) [reply config/initializers/at_exit.rb:2]: Well yes, but the docs says "Note: Cluster mode only.". This is useful both for production and development. This works also on Passenger or any other server, when process exits simply cleanup after itself and also all leftovers.

**Review by ekohl — COMMENTED** [review]
@adamruzicka I think this will also affect Dynflow which runs the Rails initializers. Could you also take a look at the potential impact?

File: config/initializers/at_exit.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  Process.getpgid(pid.to_i)
+  true
+rescue Errno::ESRCH
+  false
+end
+
+def cleanup_prometheus
+  # delete my own files
+  logmsg "Deleting #{PROMETHEUS_DIR}/*#{Process.pid}.bin"
+  FileUtils.rm_f(Dir.glob("#{PROMETHEUS_DIR}/*#{Process.pid}.bin"))
```
Comment (ekohl) [inline config/initializers/at_exit.rb]: If you're deleting `*$pid`, you will also delete unrelated files, right? Let's say you have pid 10 that exits, but pid 210 is still running. This will match both of them.

**Comment by adamruzicka** [comment]
> @adamruzicka I think this will also affect Dynflow which runs the Rails initializers. Could you also take a look at the potential impact?

If I understand this correctly:
- when telemetry is enabled with prometheus sink, the process dumps the collected metrics into `*.bin` files
- when client hits the `/metrics` endpoint, those files are read and the data is sent in the response
- when process exits, it removes its `.bin` files and any leftovers

dynflow-sidekiq processes will run this initializer so the cleanup will be done for them as well. This is fine and maybe even desired given dynflow-sidekiq processes will collect metrics and dump the bin files too

File: config/initializers/at_exit.rb (hunk 2 above)
Comment (lzap) [reply config/initializers/at_exit.rb]: Yes this is an attempt to "cleanup" this temporary directory. All the processes which are no longer active left some data which only makes things slow.

We actually do a full cleanup of this directory during startup (delete all files), so this could work fine without this extra "dead cleanup". I can remove this if needed.

**Comment by lzap** [comment]
> understand this correctly

Yes that's the gist of it.

> maybe even desired given dynflow-sidekiq processes

Yes, if those processes read YAML settings they also launch their own collectors ending up in the same directory.

**Comment by lzap** [comment]
So what you guys think?

File: config/initializers/at_exit.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+#
+# Note this is executed in a different context, Ruby on Rails application is
+# not available, this includes Rails.env, Rails.logger, Rails.root and others.
+#
+# Signal.trap cannot be used as it is redefined in web server (Puma). Multiple
+# at_exit blocks are allowed.
+#
+
+SHUTDOWN_LOGFILE = File.expand_path(File.dirname(__FILE__) + "../../../log/shutdown.log")
+PROMETHEUS_DIR = File.expand_path(File.dirname(__FILE__) + "../../../tmp/prometheus")
```
Comment (adamruzicka) [inline config/initializers/at_exit.rb]:
```suggestion
SHUTDOWN_LOGFILE = File.expand_path(__dir__ + "../../../log/shutdown.log")
PROMETHEUS_DIR = File.expand_path(__dir__ + "../../../tmp/prometheus")
```

File: config/initializers/at_exit.rb (hunk 2 above)
Comment (adamruzicka) [reply config/initializers/at_exit.rb]: +1 for removing this

**Comment by lzap** [comment]
Rebased.

**Comment by lzap** [comment]
After chat on IRC, I am improving the glob. The filename is actually `"metric_#{ metric_name }___#{ process_id }.bin"` looking in the codebase, so updated that.

File: config/initializers/at_exit.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+def logmsg(message)
+  File.open(SHUTDOWN_LOGFILE, 'a') { |f| f.puts "[#{Process.pid}] " + message }
+end
+
+def rotatelog
+  return unless File.exist?(SHUTDOWN_LOGFILE)
+  return if File.size(SHUTDOWN_LOGFILE) < 1_000_000
+  FileUtils.rm_f(SHUTDOWN_LOGFILE)
+end
```
Comment (ekohl) [inline config/initializers/at_exit.rb:22]: For production uses we already have `/etc/logrotate.d/foreman`. Would that be sufficient? It may be confusing if a small file is rotated, but part of the logs are missing. Perhaps it should just log to stdout? In production systemd will capture it and in development it may also be easier to see.

**Review by ekohl — COMMENTED** [review]
With the more precise glob I have no objections. I do wonder about shutdown.log and if we should avoid it. See inline comment.

**Comment by lzap** [comment]
After some discussion with ruby client library, it looks like this won't work. So Prometheus endpoint still has problems with too many metrics and slowing down unfortunately. I suggest to stick with statsd instead, if you need Prometheus use stats_exporter.
